### PR TITLE
[CS2113T-T10-1] SplitLah

### PR DESCRIPTION
**Pull request by froststein**
SplitLah is an application for keeping track of expenses during group outing sessions, specifically for those who are cautious about their finance. It then proceeds to split the expenses according to the group members' individual costs for the activities they participated in, for a particular session, via a Command Line Interface(CLI).

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -1,38 +1,416 @@
...
+* [Acknowledgements](#acknowledgements)
+* [Design](#design)
+    * [Architecture](#architecture)
+    * [Manager Component](#manager-component)
+    * [Profile Component](#profile-component)
+    * [TextUI Component](#textui-component)
+    * [Storage Component](#storage-component)
+    * [Parser Component](#parser-component)
+    * [Command Component](#command-component)
+* Implementation
```
Comment (laiisaac) [inline docs/DeveloperGuide.md]: Inconsistent formatting: I think there's a missing a hyperlink to the Implementation section

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -1,38 +1,416 @@
...
+* [Non-Functional Requirements](#non-functional-requirements)
+* [Glossary](#glossary)
+
 
 ## Acknowledgements
+* Address Book (Level-3) - Provide samples of User Guide(UG) and Developer Guide(DG)
+
+## Design
+### Architecture
+![Application Diagram Screenshot](https://raw.githubusercontent.com/AY2122s2-cs2113t-t10-1/tp/master/docs/images/developerguide/ArchitectureDiagram.drawio.png)
```
Comment (laiisaac) [inline docs/DeveloperGuide.md:56]: Is the sequence diagram too detailed? It is hard to see what is being written in the diagram unless it is zoomed into.
What I see at default zoom level:
image

Perhaps you can consider splitting it up into different parts (e.g. 1 diagram for loading the file, 1 diagram for processing user input).

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -1,38 +1,416 @@
...
+
+### Profile Component
+![Profile Component Screenshot](https://raw.githubusercontent.com/AY2122s2-cs2113t-t10-1/tp/master/docs/images/developerguide/Profile%20Component.drawio.png)
+<br>
+The `Profile` class holds the list of sessions and groups that have been created by the user. 
+It also tracks the unique identifier for `Session`, `Activity` and `Group` classes. The `Profile`
+class would return a unique identifier every time a new `Session`, `Activity` or `Group` is created. 
+
+### TextUI Component
+![TextUI Component Screenshot](https://raw.githubusercontent.com/AY2122s2-cs2113t-t10-1/tp/master/docs/images/developerguide/TextUI%20Component.drawio.png)
```
Comment (laiisaac) [inline docs/DeveloperGuide.md]: Screenshot for TextUI component does not load:
image

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -1,38 +1,416 @@
 # Developer Guide
```
Comment (yuzhongng) [inline docs/DeveloperGuide.md:7]: Perhaps the Developer Guide could start off with a quick introduction to what your product is about.

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -1,38 +1,416 @@
...
+   it is passed back to the `SplitLah` object to be run.
+   * In the above process, all necessary information for the execution of the command is passed to and
+     saved by the `XYZCommand` constructor.
+2. Then, `XYZCommand#run` is executed by `SplitLah`. `XYZCommand#run` carries out the task
+   that `XYZCommand` is designed to do.
+   * In general, for _data related commands_, `SessionJKLCommand`, `ActivityJKLCommand` and `GroupJKLCommand`
+     obtains the relevant `Session`, `Activity` and `Group` objects before operating upon them with relevant methods.
+   * On the other hand, for _utility commands_, `UtilityCommand` works with and uses methods from the `TextUI` component
+     to print messages and carry out their tasks.
+   * The inner workings of each of the `XYZCommand` classes can be seen in greater detail under the [Implementation section](#implementation).
```
Comment (laiisaac) [inline docs/DeveloperGuide.md:256]: Good job on the formatting as well as clarification on certain ambiguous terms used, makes it very readable for the reader

File: docs/DeveloperGuide.md (hunk 2 above)
Comment (BradenTeo) [inline docs/DeveloperGuide.md:56]: Perhaps you could label the person and the file icons on the left and right.
image

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -1,38 +1,416 @@
...
+### Profile Component
+![Profile Component Screenshot](https://raw.githubusercontent.com/AY2122s2-cs2113t-t10-1/tp/master/docs/images/developerguide/Profile%20Component.drawio.png)
+<br>
+The `Profile` class holds the list of sessions and groups that have been created by the user. 
+It also tracks the unique identifier for `Session`, `Activity` and `Group` classes. The `Profile`
+class would return a unique identifier every time a new `Session`, `Activity` or `Group` is created. 
+
+### TextUI Component
+![TextUI Component Screenshot](https://raw.githubusercontent.com/AY2122s2-cs2113t-t10-1/tp/master/docs/images/developerguide/TextUI%20Component.drawio.png)
+<br>
```
Comment (yuzhongng) [inline docs/DeveloperGuide.md:160]: The screenshot is not displayed in the Developer Guide webpage.
TextUi Component

I will assume that the screenshot is a diagram that explain how the TextUI can be changed without affecting the rest of the program.

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -1,38 +1,416 @@
...
+   of `XYZCommandParser`. This process is explained in further detail in the sequence diagrams below.
 
-{list here sources of all reused/adapted ideas, code, documentation, and third-party libraries -- include links to the original source as well}
+   ![Reference Frame Command Parser Sequence Diagram](https://raw.githubusercontent.com/AY2122s2-cs2113t-t10-1/tp/master/docs/images/developerguide/RefCommandParser.drawio.png)
 
-## Design & implementation
+   ![Reference Frame ParseABC Sequence Diagram](https://raw.githubusercontent.com/AY2122s2-cs2113t-t10-1/tp/master/docs/images/developerguide/RefParseABC.drawio.png)
 
-{Describe the design and implementation of the product. Use UML diagrams and short code snippets where applicable.}
+   ![Reference Frame InvalidCommand Instantiation Sequence Diagram](https://raw.githubusercontent.com/AY2122s2-cs2113t-t10-1/tp/master/docs/images/developerguide/RefInvalidCommand.drawio.png)
```
Comment (laiisaac) [inline docs/DeveloperGuide.md]: Diagrams provided are not explicitly linked to text explaining it. Perhaps you can consider explaining each diagram with text right below the diagram, or name the diagrams and refer to them explicitly when explaining the respective process in text.

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -1,38 +1,416 @@
...
+    * Handles UI operations of the application.
+* `Storage`
+    * Handles storage operations of the application.
+* `Parser`
+    * Handles input parsing and determines which command to run.
+* `Command`
+    * Defines how a command is executed.
+
+**Interaction between components**
+![Component Interaction Screenshot](https://raw.githubusercontent.com/AY2122s2-cs2113t-t10-1/tp/master/docs/images/developerguide/ComponentInteraction.drawio.png)
```
Comment (BradenTeo) [inline docs/DeveloperGuide.md:86]: I think it might be confusing for the user to see 2 instances of the :SplitLah class here. You could label or explain what do each of them represent. Maybe it's possible to remove one for better comprehendability.
image

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -1,38 +1,416 @@
...
+      `Parser` class, `Parser` class returns a `String[]` object to `SessionCreateCommandParser` after processing the
+      arguments.
+   * Any exception encountered by `ParserUtils` class is propagated back to `XYZCommandParser` to be handled.
+7. After all necessary information is parsed, `XYZCommandParser` instantiates a new `XYZCommand` object and passes
+   all parsed information to it through the constructor.
+8. The newly instantiated `XYZCommand` object is then returned from `XYZCommandParser` to `Parser` class,
+   and finally back to `SplitLah` to be run.
+
+### Add a session
+**API reference:** [`SessionCreateCommand.java`](https://github.com/AY2122S2-CS2113T-T10-1/tp/blob/master/src/main/java/seedu/splitlah/command/SessionCreateCommand.java)
```
Comment (laiisaac) [inline docs/DeveloperGuide.md:337]: I like that you've coloured classes and objects according to their role:
image
image
However, some diagrams such as this one does not follow the same formatting:
image
Perhaps the document would be more cohesive with consistent formatting.

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -1,38 +1,416 @@
...
+The `TextUI` class is initialized by the `Manager` class when the application starts.
+It stores a `Scanner` and `PrintStream` object supplied upon initialization to read and write to the user interface.
+It offers methods to print application output to and read user input from these objects for other classes to use.
+As TextUI handles all input and output streams, these streams can be changed without affecting the rest of the program.
+
+### Storage Component
+
+### Parser Component
+![Parser Component Screenshot](https://raw.githubusercontent.com/AY2122s2-cs2113t-t10-1/tp/master/docs/images/developerguide/ParserComponent.drawio.png)
+<br>
```
Comment (yuzhongng) [inline docs/DeveloperGuide.md:205]: Perhaps the PaserUtils could be changed to a static class
PaserUtil

File: docs/DeveloperGuide.md (hunk 8 above)
Comment (BradenTeo) [reply docs/DeveloperGuide.md:86]: Overall, I think the diagram looks quite small on the website. Perhaps can consider prioritising comprehensibility over being comprehensive.

File: docs/DeveloperGuide.md (hunk 3 above)
Comment (BradenTeo) [inline docs/DeveloperGuide.md]: The picture here cannot be seen.
image

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -1,38 +1,416 @@
...
+  `Parser` class is the only class in the `Parser` component that other external classes actively interact with.<br>
+* The `ParserUtils` class provide supporting methods for `Parser` class to properly run,
+  and `ParserErrors` class provide methods to produce custom error messages for the `Parser` component.<br>
+* The subclasses of `CommandParser` then serve to parse all arguments of a user input to create an object of a specific
+  subclass of the `Command` class.
+
+The general workflow of the `Parser` component is as follows:
+1. When required to parse for a command, the running `SplitLah` object passes a `String` object containing
+   the user input to `Parser` class.
+2. `Parser` class instantiates a new `XYZCommandParser` object corresponding to the user input
```
Comment (yuzhongng) [inline docs/DeveloperGuide.md:218]: Is the XYZCommandParser referring to the chunk of code below?
XYZCommandParser

**Review by BradenTeo — COMMENTED** [review]
Overall, I think many diagrams can be simplified and resized to be bigger so that they are visible in PDF form.

**Review by laiisaac — COMMENTED** [review]
Developer guide is generally well done, with comprehensive and correct diagrams to supplement text explanations

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -1,38 +1,460 @@
...
+### View an activity
+
+### List activities
+**API reference:** [`ActivityListCommand.java`](https://github.com/AY2122S2-CS2113T-T10-1/tp/blob/master/src/main/java/seedu/splitlah/command/ActivityListCommand.java)
+
+The sequence diagram below models the interactions between various entities in SplitLah
+when the user invokes the `activity /list` command.
+<br>
+<br>
+![List Activity Sequence Diagram Screenshot](https://raw.githubusercontent.com/AY2122s2-cs2113t-t10-1/tp/master/docs/images/developerguide/ActivityListCommand.drawio.png)
```
Comment (cczhouqi) [inline docs/DeveloperGuide.md:715]: This screenshot is also not shown.
Screen Shot 2022-03-31 at 11 11 37 AM

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -1,38 +1,460 @@
...
+4. The list of groups are stored in a `Profile` object, hence `Manager#getProfile` is called
+before the list of groups can be retrieved.
+5. The `GroupListCommand` object runs the `Profile#getGroupListSummaryString` method to retrieve a `String` object
+representing the summaries of the groups stored.
+   1. If there are no groups stored in the `Profile` object, a `String` object representing an empty list of groups is
+   returned.
+   2. Else, the `Profile` objects instantiates a new `TableFormatter` object and loops through the list of groups,
+   calling `TableFormatter#addRow` for each group to create a table with the summary of each group. A `String` object
+   representing the table is then returned.
+6. The `String` object retrieved is printed out with `TextUI#printlnMessage`.
```
Comment (cczhouqi) [inline docs/DeveloperGuide.md:902]: The diagrams and descriptions are detailed and clear. Good job!👍🏻👍🏻

File: docs/DeveloperGuide.md (hunk 2 above)
Comment (ibrahimisramos) [reply docs/DeveloperGuide.md:37]: Is the arrow from manager to textUi correct? Cause the TextUi invokes an instance the manager

**Review by cczhouqi — COMMENTED** [review]
Very detailed developer guide. Diagrams are well-designed and clear explanation is given.👍🏻

File: docs/DeveloperGuide.md (hunk 1 above)
Comment (ibrahimisramos) [reply docs/DeveloperGuide.md]: Might want to use numbers instead of bullet points given the comprehensive nature of your guide which makes very hard to follow through

File: docs/DeveloperGuide.md (hunk 2 above)
Comment (ibrahimisramos) [reply docs/DeveloperGuide.md:56]: Is the arrow from manager to textUi correct? Cause the TextUi invokes an instance the manager

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -1,38 +1,460 @@
...
+     i.e. all debts are paid and all debts are collected.
+   * A `String` object containing information regarding all transactions that have to be made is then returned.
+   * If no transactions are required to be made, a message explaining that no transactions are required to be made
+     is returned instead.
+   * For the sake of brevity, the specifics of the method `SessionSumamryCommand#processAllTransactions` is omitted
+     from the sequence diagram.
+10. Finally, with the `TextUI` object, the method `printlnMessageWithDivider` is called to print the message
+    obtained from the `SessionSummaryCommand#processAllTransactions` method.
+
+### Add an activity
```
Comment (ibrahimisramos) [inline docs/DeveloperGuide.md:555]: Do remember to include them once they are done!

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -1,38 +1,460 @@
...
+| v1.0    | Budget conscious user | create activities                    | able to track the expenditure for each activity                             |
+| v1.0    | Budget conscious user | view an existing session             | view the full details of a session previously created                       |
+| v1.0    | Budget conscious user | view an existing activity            | view the cost breakdown of who paid and who was involved in the activity    |
+| v1.0    | Budget conscious user | list all existing sessions           | view all sessions previously created                                        |
+| v1.0    | Budget conscious user | list all activities in a session     | view all the activities that happened in the session                        |
+| v1.0    | Budget conscious user | settle all transactions of a session | see a summary of who needs to pay what amount to who for the entire session |
+| v1.0    | User                  | exit the application                 | stop tracking                                                               |
```
Comment (ibrahimisramos) [inline docs/DeveloperGuide.md]: Do add in v2.0 and can consider having a separate table for each user type or sort them by different types of user

File: docs/DeveloperGuide.md [hunk 16]
```diff
@@ -1,38 +1,460 @@
...
 
 ## Non-Functional Requirements
-
-{Give non-functional requirements}
+1. The application should be able to work in any operating systems with `Java 11` installed.
+2.
```
Comment (ibrahimisramos) [inline docs/DeveloperGuide.md]: Otherwise good job! Very detailed but might be a bit too detailed

**Review by ibrahimisramos — COMMENTED** [review]
Great job team!
